File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        
        #empty
        if not head or not head.next:
            return True
        
        #find mid point
        fast = head
        slow = head
        while fast != None and fast.next != None:
            slow = slow.next
            fast = fast.next.next

        #reverse first half
        prev = None
        curr = head
        while curr != slow:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt


        #handles odd-length (skip middle)
        #if Fast is None, odd list.
        if fast:
            slow = slow.next

        #compare
        left = prev
        right = slow 
        while left and right != None:
            if left.val != right.val:
                return False
            left = left.next
            right = right.next
        return True





        

        
        
        

        

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        
        //1. We gotta find the half way point.
        //2. Reverse half the list.
        //3. Check palindrome


        //ListNode* curr = head;
        ListNode* slow = head;
        ListNode* fast = head;
        ListNode* prev = nullptr;
        ListNode* nxt = nullptr;

        //Even
        while (fast && fast->next)
        {
            //Moves our fast ptr..
            fast = fast->next->next;

            //Reverse linkedlist
            nxt = slow->next;
            slow->next = prev;
            prev = slow;
            slow = nxt;
        }

        //Odd??
        if (fast)
        {
            slow = slow->next;
        }

        //Palindrome...
        while (prev && slow)
        {
            if (prev->val != slow->val)
            {
                return false;
            }

            prev = prev->next;
            slow = slow->next;
        }

        return true;
    }
};